### PR TITLE
Implement a logging callback that unifies multiple loggers

### PR DESCRIPTION
**Pull request by ibro45**
Aims to implement a LighterLogger as a PL Callback rather than a PL Logger, as the latter's implementation is very messy.

### CODE REVIEW

**Comment by ibro45** [comment]
Currently support for logging to Tensorboard and Wandb:

- loss
- metrics
- input, pred, target data. Either as an image or a single scalar.

What's next:

- support for other input, pred, target data types. E.g. logging strings, multiple scalars, histograms etc. This will need projects that deal with such cases. We should tackle that as we encounter such projects IMO. Will create an issue for it.
- config (hparams) logging. This needs the YAML config of the run to be dumped before the instantiation of the logger. Probably will be dumped by `cli.py` into a temp file and then moved into the `log_dir` of the Logger and logged to tensorboard, wandb, etc. This will be a separate PR, opening an issue for this too.
- implement other loggers. MLflow or other, the idea is to have them all, but again, something to add as we start using them as this is not a priority right now.
- In DDP, each rank should log its own log file. This will require getting all the `logging.Loggers` of all libraries (e.g. Monai or PyTorch Lightning) to work with our `loguru.Logger` so that everything can be logged properly. Will create an issue.

Ready for review.

**Review by kbressem — COMMENTED** [review]
I overall like it and think it is well thought through. Only think I am unsure about is the frequent use of `sys.exit()` instead of raising errors. This makes try/except impossible, does not allow for Tracebacks and makes testing hard. 

Tests is also something I would recommend for this PR. At least initialization and basic mock of logging to file.

**Comment by surajpaib** [comment]
@ibro45 and I had a few discussions about the `sys.exit` strategy before. 

In the `logger.py`, I think since all of the `sys.exit()` are very well-defined error cases, having tracebacks might not even be necessary imo. 

However, @ibro45 the ability to get defined values returned from the code might be very useful especially for testing when we can build test cases accordingly. Maybe its a good idea to have a more consistent standard across the entire library that allows handling both well-defined and ill-defined error cases.

File: lighter/callbacks/logger.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+                    # On rank 0, sum the losses from all ranks. Other ranks remain with the same loss as before.
+                    dist.reduce(loss, dst=0)
+                    # On rank 0, average the loss sum by dividing it with the number of processes.
+                    loss = loss.item() / dist.get_world_size() if dist.get_rank() == 0 else loss.item()
+                # Divide the accumulated loss by the number of steps in the epoch.
+                loss /= self.epoch_step_counter[mode]
+                outputs["loss"] = loss
+            
+            # Metrics
+            # Get the torchmetrics.
```
[the remark this answers is not in the document]
Comment (surajpaib) [reply lighter/callbacks/logger.py:238]: I think its a good idea also to have other metrics supported. But maybe a separate issue for later?

**Comment by surajpaib** [comment]
@ibro Testing the branch right now, are the `.log` files expected to be empty?

File: lighter/callbacks/logger.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+from yaml import safe_load
+
+from lighter import LighterSystem
+
+LIGHTNING_TO_LIGHTER_STAGE = {"train": "train", "validate": "val", "test": "test"}
+OPTIONAL_IMPORTS = {}
+
+class LighterLogger(Callback):
+
+    def __init__(self, project, log_dir, log_input_as=None, log_target_as=None, log_pred_as=None, tensorboard=False, wandb=False) -> None:
```
Comment (surajpaib) [inline lighter/callbacks/logger.py:23]: Could we rename this to `log_input_as` to `input_type` and the others respectively? I think it `log` is unnecessary as its in the logger callback already

File: lighter/callbacks/logger.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+    def on_validation_epoch_end(self, trainer: Trainer, pl_module: LighterSystem) -> None:
+        self._on_epoch_end(trainer, pl_module)
+
+    def on_test_epoch_end(self, trainer: Trainer, pl_module: LighterSystem) -> None:
+        self._on_epoch_end(trainer, pl_module)
+
+
+def preprocess_image(image: torch.Tensor) -> torch.Tensor:
+    """Preprocess the image before logging it. If it is a batch of multiple images,
```
[the remark this answers is not in the document]
Comment (surajpaib) [reply lighter/callbacks/logger.py:301]: The logger actually does this now when passed with `log_input_as: image_N` where `N` is the number of images to be pushed. 

@ibro45 It might be a bit hard to explain this feature to general users in this format. Could we have something like `max_samples` that applies to input, target and pred together? Ties in with the `input_type` comment.

File: lighter/callbacks/logger.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+        # Running loss. Resets at each epoch. Loss is not calculated for `test` mode.
+        self.loss = {"train": 0, "val": 0}
+        # Epoch steps. Resets at each epoch. No `test` mode step counter as loss is not calculated for it.
+        self.epoch_step_counter = {"train": 0, "val": 0}
+        # Global steps. Replaces `Trainer.global_step` because it counts optimizer steps
+        # instead of batch steps, which can be problematic when using gradient accumulation.
+        self.global_step_counter = {"train": 0, "val": 0, "test": 0}
+
+    def setup(self, trainer: Trainer, pl_module: LighterSystem, stage: str) -> None:
```
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:41]: Unfortunately not possible as PyTorch Lightning automatically passes the argument with that name. I type hinted is as a LighterSystem (which still is a LightningModule) because this logger requires some parts that are specific to LighterSystem, for example: https://github.com/project-lighter/lighter/blob/30b15fd41b5c821230dc5141020e57b3dbb68ec5/lighter/callbacks/logger.py#L239

File: lighter/callbacks/logger.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+            tensorboard_dir.mkdir()
+            self.tensorboard = tensorboard.SummaryWriter(log_dir=tensorboard_dir)
+            # self.tensorboard.add_hparams(config)
+
+        # Wandb initialization.
+        if self.wandb:
+            OPTIONAL_IMPORTS["wandb"], wandb_available = optional_import("wandb")
+            if not wandb_available:
+                logger.error("W&B not installed. To install it, run `pip install wandb`. Exiting.")
+                sys.exit()
```
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:78]: It's not required, but the user specifies if they want to use wandb:
https://github.com/project-lighter/lighter/blob/30b15fd41b5c821230dc5141020e57b3dbb68ec5/lighter/callbacks/logger.py#L23

So if they want to use it, but it's not installed, it should exit. If they don't want to use it, they they set `wandb=False`

File: lighter/callbacks/logger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:238]: Metrics have to be `torchmetrics`:
https://github.com/project-lighter/lighter/blob/f7ac3d6260c43894a93fcb8f0d755bbed90241de/lighter/system.py#L70

 I wouldn't support any other format of metric. I think that accounting for all the possible ways of doing it would quickly create a mess here:
  https://github.com/project-lighter/lighter/blob/f7ac3d6260c43894a93fcb8f0d755bbed90241de/lighter/callbacks/logger.py#L242
  
 If a user wants a metric that's not available in `torchmetrics` they should implement it as a `torchmetrics.Metric` themself. It's pretty easy https://torchmetrics.readthedocs.io/en/stable/pages/implement.html.

However,  from a quick glance it appears that Monai metrics have a unified API too - we could write a metric wrapper that turns a Monai metric into a TorchMetric metric. It'd go to our `contrib` repo.

What do you guys think?

File: lighter/callbacks/logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:301]: Agree with max_samples, and it can be used for other data types too, not only images. Also agree that `image_N` might be unnecessarily complicated.

Do you think you'd want to have `max_samples` different for input, target, and pred? I can't think of a case where I would, but not sure.

**Comment by ibro45** [comment]
@surajpaib @kbressem both of you brought up `sys.exit()` independently, pretty convinced now that we should change it. That would be a separate PR though as I'll have to go through the whole repo replacing them.

Any preferences/guidelines on how we should handle them?

File: lighter/callbacks/logger.py (hunk 2 above)
Comment (ibro45) [reply lighter/callbacks/logger.py:23]: Sounds good, will do

File: lighter/callbacks/logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (surajpaib) [reply lighter/callbacks/logger.py:301]: I can't think of any to be honest, especially when logging, you'd want all to be consistent I'd think.

File: lighter/callbacks/logger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (surajpaib) [reply lighter/callbacks/logger.py:238]: That sounds perfect then! Since `torchmetrics` handle so many things internally, I agree that this is the best approach

**Comment by ibro45** [comment]
@surajpaib 
> @Ibro Testing the branch right now, are the .log files expected to be empty?

That one was confusing for me too, I remember it logging it first and then it wouldn't. Can you check if the file was populated with some logs at the end?

The logs only contain the Lighter logs though. No logs from other libraries. Tried to intercept them all into our loguru Logger, but it didn't work correctly with DDP - all ranks ended up logging at the same time. I'd like to resolve this issue together with DDP per-rank log files. Probably best to comment it out for the moment. What do you think?

**Comment by surajpaib** [comment]
> @surajpaib
> 
> > @Ibro Testing the branch right now, are the .log files expected to be empty?
> 
> That one was confusing for me too, I remember it logging it first and then it wouldn't. Can you check if the file was populated with some logs at the end?
> 
> The logs only contain the Lighter logs though. No logs from other libraries. Tried to intercept them all into our loguru Logger, but it didn't work correctly with DDP - all ranks ended up logging at the same time. I'd like to resolve this issue together with DDP per-rank log files. Probably best to comment it out for the moment. What do you think?

Works for me, lets not have this atm and do it properly all together. Including logging for raising errors and exceptions

**Comment by surajpaib** [comment]
Great work overall with this! Checked the logs on `wandb`, looks very nice!

File: lighter/callbacks/logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:299]: Implemented, it cleaned up the code nicely too

**Comment by ibro45** [comment]
Don't mind me not formatting the code, I decided not to do it as it'll be done by Suraj's CI anyway after it's merged.

Also, I'll write a docstring for the `__init__` when this PR is ready to merge.

**Review by surajpaib — APPROVED** [review]

File: lighter/callbacks/logger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kbressem) [reply lighter/callbacks/logger.py:236]: Ok, I agree. But then the logger should raise an Exception, if a metric is passed, that is not a `torchmetrics.Metric`. Given this PR is merged, shall I raise an issue for this?

[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:236]: The check would be in the LighterSystem. But if we're doing it for metrics, we should then type-check everything:
https://github.com/project-lighter/lighter/blob/08cb305e965b9434411c59b752db40c454a5dc40/lighter/system.py#L20

That becomes quite a hassle, I guess this is when I wish there was runtime type checking in Python.

**Comment by ibro45** [comment]
> I would prefer to raise Exception, maybe even custom Exceptions. This has the advantage, that you can test for the code if it raises correctly.

I'll have to read up on what's the most recommended way to do it.

> ... but using unittest would make integration to monai easier, in case we want to push some of our stuff to their library.

I don't think that anything in `lighter` core will be going in the `monai` direction, as `lighter` is meant for general purposes. You can do `monai`-related projects, but also something else completely. We rely on their config system mainly because it's amazing, not because we want it to be completely tied to `monai`. The fact that, as a result of it, we can bundle our \*medical imaging projects\* easily just sweetens the deal (and hopefully we'll have many projects that rely on that). But, the goal of Lighter is to be as agnostic as possible, enabling you to work with all kinds of data.

That said, I'm sure that some of the `lighter.contrib` will go into that direction, as we'll have parts that are related to Monai, like custom Monai transforms.

Regarding pytest vs unittest, I have much to learn about testing, so I have no preference or opinion.

File: lighter/callbacks/logger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibro45) [reply lighter/callbacks/logger.py:236]: @surajpaib wdy think?
